assets/src/edit-story/components/form/numeric.js

```diff
@@ -97,48 +117,195 @@ function Numeric({
   canBeEmpty,
   ...rest
 }) {
-  const isMultiple = value === MULTIPLE_VALUE;
-  const placeholder = isMultiple ? __('multiple', 'web-stories') : '';
-  const [dot, setDot] = useState(false);
   const [inputValue, setInputValue] = useState(value);
-  const ref = useRef();
 
-  useEffect(() => setInputValue(value), [value]);
+  const inputRef = useRef();
+  const targetFormRef = useRef(null);
+  const skipValidationRef = useRef(false);
+  const selectInputContents = useRef(false);
+  const selectContentsTimeout = useRef(-1);
+  const submitOnChangeTimeout = useRef(-1);
+
+  const { focused, handleFocus, handleBlur } = useFocusAndSelect(inputRef);
+
+  const isMultiple = inputValue === MULTIPLE_VALUE;
+  const placeholder = isMultiple ? __('multiple', 'web-stories') : '';
+
+  const handleChange = useCallback((event) => {
+    setInputValue(event.target.value);
+  }, []);
+
+  const submitOnChange = useCallback(
+    (val, { selectContentOnUpdate }) => {
+      selectInputContents.current = selectContentOnUpdate;
+      onChange(val);
+    },
+    [onChange]
+  );
+
+  const validateAndSubmitInput = useCallback(
+    (
+      val,
+      { ignoreValidation, selectContentOnUpdate, debounceOnChange } = {}
+    ) => {
+      targetFormRef.current = inputRef.current.form;
+
+      const validValue = ignoreValidation
+        ? val
+        : validateInput(val, {
+            canBeNegative,
+            canBeEmpty,
+            float,
+            defaultValue: value,
+          });
+
+      setInputValue(validValue);
+
+      window.clearTimeout(submitOnChangeTimeout.current);
+      if (debounceOnChange) {
+        submitOnChangeTimeout.current = window.setTimeout(() => {
+          submitOnChange(validValue, { selectContentOnUpdate });
+        }, ONCHANGE_DEBOUNCE_DELAY);
+      } else {
+        submitOnChange(validValue, { selectContentOnUpdate });
+      }
```

Comment:
@miina I added debounce logic so when a user uses the up/down arrow to quickly increment the value, instead of sending a complete `onChange` event per fast click (which is what caused the performance lag), it'll wait until the user stops rapid fire clicking and then send the last change down the pipeline.

Reply:
Works well now!

One note: should the timeout be canceled if it exists and the user types in something? If that's possible. Otherwise it might theoretically happen that the user uses arrows, and then types in something else, but due to debouncing, the "arrowed" value will be submitted and change the input value in the middle. Thoughts?

Reply:
You're right!  It should be cancelled every time the user types in anything along with any time they submit a new value.  I'll add a `clearTimeout` call on the `handleChange` callback which triggers every time a user types in a new character in the input field (but doesn't trigger when they just hit the up/down arrows).  Nice catch!